File: 16 - 3Sum Closest/solution.py
import math
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        if len(nums) == 3:
            return sum(nums)
        
        nums.sort()

        final = len(nums)
        
        minimum_diff = math.inf

        closest_sum = math.inf

        for i in range(len(nums) - 2):
            left, right = i + 1, len(nums) - 1

            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                current_diff = abs(current_sum - target)

                if current_diff < minimum_diff:
                    minimum_diff = current_diff
                    closest_sum = current_sum

                if current_sum < target:
                    left += 1
                else:
                    right -= 1
        
        return closest_sum


if __name__ == '__main__':
    solution = Solution()
    print(solution.threeSumClosest([-1,2,1,-4], 1))

        

File: 16 - 3Sum Closest/solution.c
#include <stdio.h>
#include <stdlib.h>

void quicksort(int* nums, int left, int right) {
    if (left >= right) return;
    int pivot = nums[right];
    int i = left, j;
    for (j = left; j < right; j++) {
        if (nums[j] < pivot) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
            i++;
        }
    }
    int temp = nums[i];
    nums[i] = nums[right];
    nums[right] = temp;
    quicksort(nums, left, i - 1);
    quicksort(nums, i + 1, right);
}

int threeSumClosest(int* nums, int numsSize, int target) {
    int left;
    int right;
    int currentSum;
    int currentDiff;
    int minimumDiff = __INT16_MAX__;
    int closestSum = __INT16_MAX__;
    
    if (numsSize == 3) {
        return nums[0] + nums[1] + nums[2];
    }

    quicksort(nums, 0, numsSize -1);

    for (int i = 0; i < numsSize -2; i++) {
        left = i + 1;
        right = numsSize - 1;

        while (left < right) {
            currentSum = nums[i] + nums[left] + nums[right];
            currentDiff = abs(currentSum - target);

            if (currentDiff < minimumDiff) {
                minimumDiff = currentDiff;
                closestSum = currentSum;
            }

            if (currentSum < target) {
                left++;
            } else {
                right--;
            }
        }
    }
    return closestSum;
}

int main() {
    int nums[] = {-1, 2, 1, -4};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int target = 1;
    int result = threeSumClosest(nums, numsSize, target);
    printf("Output: %d\n", result); // Esperado: 2
    return 0;
}